### DISCUSSION TITLE
Support references before definitions in DSL

### DISCUSSION THREAD

**Discussion by mattdrees** [structurizr/java#329] [Q&A] [answered]
### Description

It looks like the parser does a single pass through the DSL file, and validates references as it goes.

So, this is valid:
```
workspace {
  model {
        a = softwareSystem "A"
        b = softwareSystem "B"
        a -> b
  }
}
```

but this is not:
```
workspace {
  model {
        a = softwareSystem "A"
        a -> b
        b = softwareSystem "B"
  }
}
```
This is the error message from `structurizr-cli validate`:

> The destination element "b" does not exist at line 4 of /myprojects/structurizr-java/structurizr-dsl/src/test/resources/dsl/relationship-ref-before-target.dsl: a -> b

Not the end of the world, but I've several times wished I could define a relationship in or around a source system (system "A"), rather than having to define it way down in the file after the target system (system "B").

Does that sound like something you'd be interested in supporting?

### Priority

I have no budget and there's no rush, please add this feature for free

**Comment by simonbrowndotje** [comment, marked as answer]
This has been discussed a few times (example -> https://github.com/structurizr/java/discussions/260) ... it's not something that's particularly viable I'm afraid.

**Comment by mattdrees** [comment]
I see, thanks. Unfortunate.

It feels like it wouldn't be very hard to move relationship processing 'to the end'. But I'm inferring the main downside would be that scripts and plugins wouldn't see relationships 'above them' when they run. Is that about right?

**Reply by simonbrowndotje** [reply]
> It feels like it wouldn't be very hard to move relationship processing 'to the end'.

It does feel like that, yes. And indeed that would work for basic single line definitions like this:

```
workspace {

    model {
        a = softwareSystem "A"
        a -> b

        b = softwareSystem "B"
    }

}
``` 

But it wouldn't work for something like this:

```
workspace {

    !identifiers hierarchical

    model {
        a = softwareSystem "A" {
            aa = container "AA"
            
            aa -> b.bb {
                tags "Tag 1"
                perspectives {
                    "Name" "Value"
                }
            }
        }

        b = softwareSystem "B" {
            bb = container "BB"
        }
    }

}
```

`!identifiers hierarchical` forces you to reference elements by their full name, except when the element is available inside the current scope - so `aa -> b.bb` works where it is, but would need to become `a.aa -> b.bb` if moved outside of the parent software system. Additionally, you'd need to also move all of the relationship tags and perspectives definitions.

As I've said to everybody who has asked about this ... the code is open source, so you're more than welcome to fork it and/or submit a PR. Alternatively, you might want to look at LikeC4, which provides this feature -> https://likec4.dev/dsl/references/#hoisting

**Reply by mattdrees** [reply]
Thanks for the thoughtful response! 

I will see if I can find a reasonable way to do this in the code.
